pkg/distributor/distributor.go

```diff
@@ -326,7 +326,7 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 			if sp := opentracing.SpanFromContext(ctx); sp != nil {
 				localCtx = opentracing.ContextWithSpan(localCtx, sp)
 			}
-			d.sendSamples(localCtx, ingester, samples, &tracker)
+			d.sendSamples(localCtx, req.PushID, ingester, samples, &tracker)
```

Comment:
Are we not going to have a problem if multiple streams are created with the same pushID and end up on the same replica set ?

Reply:
But this pushID is unique per batch. 

Unless the same exact batch is retried, ingesters will never see the same pushID.

I'm failing to understand how mutiple streams in the batch can cause problem here. Am i missing something?

Reply:
@kavirajk I think I see what Cyril is saying. We could have a lot of series in the push request, and it's likely that some might have overlapping ingester replica sets. Here's how I understand the code you've written at the moment.

 - we start by splitting a push request into individual streams then validating each streams labels and entries (line 232 to 282)
- at the end of the loop for the above we append the key (user id + stream labels) for the stream as well as the stream tracker (which contains just this stream)
- next we loop over the keys and find the ingester replica set for that key, then we append to a slice of samples for each ingester by looping over the set of instances in the replica set (line 300 to 314)
- finally we loop over the list of 'samples for each ingester' and send the samples (stream tracker) to the ingester, using the original push requests push ID as the idempotent identifier (line 321 to 332)

I think the way we can end up rejecting valid Pushes is as follows;
- we have a push request come into the distributor with a request ID of 1234
- this push request has a sample for "series A" and a sample for "series B"
- the replica set for "series A" resolves to ingesters 0-2
- the replica set for "series B" resolves to ingesters 1-3
- the push request generated for "series A" to be sent to ingesters 0-2 has an 'idempotent identifier' of 1234
- the push request generated for "series B" to be sent to ingesters 1-3 also has an 'idempotent identifier' of 1234

So depending on which push request reaches each ingester first they will accept and reject some samples for **different** series from the same original push request.